### PR TITLE
Xamarin functional tests for setUpdateTrack

### PR DESCRIPTION
**Pull request by russelarms**
Things to consider before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [ ] Did you add unit tests if this modifies the Windows code?
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

Xamarin functional tests for setUpdateTrack.

## Related PRs or issues

[AB#75371](https://msmobilecenter.visualstudio.com/Mobile-Center/_workitems/edit/75371)

### CODE REVIEW

File: Tests/Contoso.Test.Functional.iOS/AppDelegate.cs [hunk 1]
```diff
@@ -12,6 +12,8 @@
 
 namespace Contoso.Test.Functional.iOS
 {
+
+    using iOSKeyChainUtil = Microsoft.AppCenter.iOS.Bindings.MSKeychainUtil;
```
Comment (MatkovIvan) [inline Tests/Contoso.Test.Functional.iOS/AppDelegate.cs]:
```suggestion
    using iOSKeyChainUtil = Microsoft.AppCenter.iOS.Bindings.MSKeychainUtil;
    
```

File: Tests/Microsoft.AppCenter.Test.Functional/ExpectedData.cs [hunk 2]
```diff
@@ -11,5 +12,6 @@ internal class ExpectedData
         internal HttpResponse Response { get; set; }
         internal Func<RequestData, bool> Where { get; set; }
         internal TaskCompletionSource<RequestData> TaskCompletionSource { get; set; }
+        internal CancellationTokenSource cancellationTokenSource { get; set; }
```
Comment (MatkovIvan) [inline Tests/Microsoft.AppCenter.Test.Functional/ExpectedData.cs]:
```suggestion
        internal CancellationTokenSource СancellationTokenSource { get; set; }
```

File: Tests/Microsoft.AppCenter.Test.Functional/HttpNetworkAdapter.cs [hunk 3]
```diff
@@ -28,13 +28,20 @@ public Task<RequestData> MockRequestByLogType(string logType, HttpResponse respo
...
+                TaskCompletionSource = new TaskCompletionSource<RequestData>()
             };
+
+            // Since we can't directly assign token to a Task created with FromResult,
+            // and we can't work with token in SendAsync because it may or may not be called after this method,
+            // we are registering anonymous cancellation token here.
+            // The only weak point can occur if there's a significant delay between actual Task creation time and calling this method,
+            // but since we only use that in tests, that can be ignored.
+            expectedData.cancellationTokenSource = new CancellationTokenSource(TimeSpan.FromSeconds(delayTimeInSeconds));
+            expectedData.cancellationTokenSource.Token.Register(() => expectedData.TaskCompletionSource.TrySetCanceled());
```
Comment (MatkovIvan) [inline Tests/Microsoft.AppCenter.Test.Functional/HttpNetworkAdapter.cs]: Can we move it into `ExpectedData` (+ Dispose)?

**Comment by Jamminroot** [comment]
Please note that this PR is intended to work with #1312 changes, so should probably be merged only after 1312 merged here, or, if base branch is changed, along with #1312 to a new base branch.

File: Tests/Microsoft.AppCenter.Test.Functional/ExpectedData.cs (hunk 2 above)
Comment (Jamminroot) [reply Tests/Microsoft.AppCenter.Test.Functional/ExpectedData.cs]: Fixed in separate [PR's commit](https://github.com/microsoft/appcenter-sdk-dotnet/pull/1312/commits/35970844bebf672221b1e338b41c1afa079976e8)

File: Tests/Microsoft.AppCenter.Test.Functional/HttpNetworkAdapter.cs (hunk 3 above)
Comment (Jamminroot) [reply Tests/Microsoft.AppCenter.Test.Functional/HttpNetworkAdapter.cs]: Resolved in separate [PR's commit](https://github.com/microsoft/appcenter-sdk-dotnet/pull/1312/commits/35970844bebf672221b1e338b41c1afa079976e8)

File: Tests/Contoso.Test.Functional.iOS/AppDelegate.cs [hunk 4]
```diff
@@ -22,7 +24,7 @@ public class AppDelegate : RunnerAppDelegate
 
         private static UIApplication UiApplication;
         private static NSDictionary LaunchOptions;
-
+
```
Comment (russelarms) [inline Tests/Contoso.Test.Functional.iOS/AppDelegate.cs]:
```suggestion

```

File: Tests/Contoso.Test.Functional.iOS/AppDelegate.cs [hunk 5]
```diff
@@ -22,7 +25,6 @@ public class AppDelegate : RunnerAppDelegate
 
         private static UIApplication UiApplication;
         private static NSDictionary LaunchOptions;
-
```
Comment (MatkovIvan) [inline Tests/Contoso.Test.Functional.iOS/AppDelegate.cs]: Revert this empty line

File: Tests/Microsoft.AppCenter.Test.Functional/Distribute/DistributeUpdateTest.cs [hunk 6]
```diff
@@ -28,11 +30,94 @@ public class DistributeUpdateTest
...
+            var eventTask = httpNetworkAdapter.MockRequest(request => request.Method == "GET");
+            var startServiceTask = httpNetworkAdapter.MockRequestByLogType("startService");
+
+            // Start AppCenter.
+            AppCenter.UnsetInstance();
+            AppCenter.LogLevel = LogLevel.Verbose;
+            AppCenter.Start(Config.ResolveAppSecret(), typeof(Distribute));
+
+            // Wait for "startService" log to be sent.
+            Task.WaitAny(startServiceTask, Task.Delay(5000));
```
Comment (MatkovIvan) [inline Tests/Microsoft.AppCenter.Test.Functional/Distribute/DistributeUpdateTest.cs]: It doesn't fail on timeout. Use timeout logic from adapter.

File: SDK/AppCenterDistribute/Microsoft.AppCenter.Distribute.Android/Distribute.cs [hunk 7]
```diff
@@ -36,6 +36,11 @@ static void PlatformSetApiUrl(string apiUrl)
             AndroidDistribute.SetApiUrl(apiUrl);
         }
 
+        internal static void PlatformUnsetInstance()
```
Comment (guperrot) [inline SDK/AppCenterDistribute/Microsoft.AppCenter.Distribute.Android/Distribute.cs]: private

File: SDK/AppCenterDistribute/Microsoft.AppCenter.Distribute.Shared/Distribute.cs [hunk 8]
```diff
@@ -83,5 +83,10 @@ public static UpdateTrack UpdateTrack
                 SetUpdateTrack(value);
             }
         }
+
+        public static void UnsetInstance()
```
Comment (guperrot) [inline SDK/AppCenterDistribute/Microsoft.AppCenter.Distribute.Shared/Distribute.cs]: internal

File: SDK/AppCenterDistribute/Microsoft.AppCenter.Distribute.iOS/Distribute.cs [hunk 9]
```diff
@@ -112,6 +112,11 @@ static UpdateTrack GetUpdateTrack()
             return (UpdateTrack)updateTrackValue;
         }
 
+        internal static void PlatformUnsetInstance()
```
Comment (guperrot) [inline SDK/AppCenterDistribute/Microsoft.AppCenter.Distribute.iOS/Distribute.cs]: private

File: SDK/AppCenterDistribute/Microsoft.AppCenter.Distribute/Distribute.cs [hunk 10]
```diff
@@ -41,5 +41,9 @@ static UpdateTrack GetUpdateTrack()
         {
             return UpdateTrack.Public;
         }
+
+        internal static void PlatformUnsetInstance()
```
Comment (guperrot) [inline SDK/AppCenterDistribute/Microsoft.AppCenter.Distribute/Distribute.cs]: private

File: Tests/Contoso.Test.Functional.iOS/AppDelegate.cs [hunk 11]
```diff
@@ -12,6 +12,9 @@
 
 namespace Contoso.Test.Functional.iOS
 {
+
```
Comment (guperrot) [inline Tests/Contoso.Test.Functional.iOS/AppDelegate.cs]:
```suggestion
```

File: Tests/Contoso.Test.Functional.iOS/AppDelegate.cs [hunk 12]
```diff
@@ -22,7 +25,7 @@ public class AppDelegate : RunnerAppDelegate
 
         private static UIApplication UiApplication;
         private static NSDictionary LaunchOptions;
-
+
```
Comment (guperrot) [inline Tests/Contoso.Test.Functional.iOS/AppDelegate.cs]:
```suggestion

```

File: Tests/Microsoft.AppCenter.Test.Functional/Distribute/DistributeUpdateTest.cs [hunk 13]
```diff
@@ -28,11 +30,94 @@ public class DistributeUpdateTest
...
+            var startServiceTask = httpNetworkAdapter.MockRequestByLogType("startService");
+
+            // Start AppCenter.
+            AppCenter.UnsetInstance();
+            AppCenter.LogLevel = LogLevel.Verbose;
+            AppCenter.Start(Config.ResolveAppSecret(), typeof(Distribute));
+
+            // Wait for "startService" log to be sent.
+            await startServiceTask;
+
```
Comment (guperrot) [inline Tests/Microsoft.AppCenter.Test.Functional/Distribute/DistributeUpdateTest.cs]: Add a comment

File: Tests/Microsoft.AppCenter.Test.Functional/Distribute/DistributeUpdateTest.cs [hunk 14]
```diff
@@ -28,11 +30,94 @@ public class DistributeUpdateTest
...
+            // Start AppCenter.
+            AppCenter.UnsetInstance();
+            AppCenter.LogLevel = LogLevel.Verbose;
+            Distribute.UpdateTrack = UpdateTrack.Public;
+            AppCenter.Start(Config.ResolveAppSecret(), typeof(Distribute));
+
+            // Wait for "startService" log to be sent.
+            await startServiceTask;
+
+            DistributeEvent?.Invoke(this, DistributeTestType.OnResumeActivity);
```
Comment (guperrot) [inline Tests/Microsoft.AppCenter.Test.Functional/Distribute/DistributeUpdateTest.cs]: Add a comment (or remove blank line)

**Pull request merged**